pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -186,25 +241,57 @@ func (l *ListOptionIndexer) afterDelete(key string, tx db.TXClient) error {
 	return nil
 }
 
+func (l *ListOptionIndexer) deleteLabels(key string, tx db.TXClient) error {
+	err := tx.StmtExec(tx.Stmt(l.deleteLabelsStmt), key)
+	if err != nil {
+		return &db.QueryError{QueryString: l.deleteLabelsQuery, Err: err}
+	}
+	return nil
+}
+
+type QueryInfo struct {
+	query       string
+	params      []any
+	countQuery  string
+	countParams []any
+	limit       int
+	offset      int
+}
+
 // ListByOptions returns objects according to the specified list options and partitions.
 // Specifically:
 //   - an unstructured list of resources belonging to any of the specified partitions
 //   - the total number of resources (returned list might be a subset depending on pagination options in lo)
 //   - a continue token, if there are more pages after the returned one
 //   - an error instead of all of the above if anything went wrong
 func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, partitions []partition.Partition, namespace string) (*unstructured.UnstructuredList, int, string, error) {
-	// 1- Intro: SELECT and JOIN clauses
-	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, db.Sanitize(l.GetName()))
+	queryInfo, err := l.constructQuery(lo, partitions, namespace, db.Sanitize(l.GetName()))
+	if err != nil {
+		return nil, 0, "", err
+	}
+	return l.executeQuery(ctx, queryInfo)
+}
+
+func (l *ListOptionIndexer) constructQuery(lo ListOptions, partitions []partition.Partition, namespace string, dbName string) (*QueryInfo, error) {
+	queryInfo := &QueryInfo{}
+
+	// 1- First, what kind of filtering will be doing?
+	// 1.1- Intro: SELECT and JOIN clauses
+	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, dbName)
 	query += "\n  "
-	query += fmt.Sprintf(`JOIN "%s_fields" f ON o.key = f.key`, db.Sanitize(l.GetName()))
+	query += fmt.Sprintf(`JOIN "%s_fields" f ON o.key = f.key`, dbName)
+	if hasLabelFilter(lo.Filters) {
+		query += "\n  "
+		query += fmt.Sprintf(`JOIN "%s_labels" lt ON o.key = lt.key`, dbName)
```

Comment:
I fear this will not always work.

Main failure case will be when a single object row has multiple labels: `JOIN` will create multiple resulting rows with all object/field columns repeated, and different label columns.

Secondarily, if an object has no labels at all then it will be filtered out by this `JOIN`.


One way to resolve this problem could be to not JOIN the labels table at all here, and only use subqueries in filters when needed.

Another is to `LEFT OUTER JOIN` here, and then deduplicate by DISTINCT on all other columns before returning results. Probably more convoluted.


Do we have any tests involving the database that cover this?

Reply:
My main peeve here is that we're mocking the database in the unit tests, and that makes increasing less sense to me.

I've written plenty of unit tests in Rails and Flask (python), and you never mock the database.  You just use a test variant of it, and the framework resets the database on every test. Yes, it takes a bit longer, but the tests are much simpler (no mocks) and you have more confidence that the DB code will run correctly in production.

So, no, we don't have any tests that cover this. 

Yes, I'm proposing that we rewrite the unit tests .

In my defense: https://markphelps.me/posts/writing-tests-for-your-database-code-in-go/ : 
```
However, I would argue that mocking the database when testing your SQL code is an anti-pattern that you should avoid, mainly because it violates requirement 3[*] since it isn’t actually testing that we can correctly interact with a real database as we are mocking the database itself!
```
* "We want to test that we can correctly interact with the database and also test that the business logic is correct."

I certainly worked out my statements interactively with a sqlite DB, namely the one created by running steve while creating a few events, and then shutting down steve so the DB wouldn't change.

Here's an example:
```
sqlite> select * from _v1_Event_labels;
default/my-custom-event4|app|app1
default/my-custom-event4|bar|chocolate
default/my-custom-event5|app|app1
default/my-custom-event5|bar|chocolate
default/my-custom-event5|cat|morris
default/my-custom-event5|dog|zipher
sqlite> SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o JOIN "_v1_Event_fields" f on o.key = f.key JOIN "_v1_Event_labels" lt on o.key = lt.key WHERE (lt.label = "dog" and lt.value = "zipher");
%
 Unstructured??||0
```

This looks ok. Unfortunately the fields we're asking for our partly binary and not very useful.  But let's look at a label that is used more than once:
```
sqlite> SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o JOIN "_v1_Event_fields" f on o.key = f.key JOIN "_v1_Event_labels" lt on o.key = lt.key WHERE (lt.label = "bar" and lt.value = "chocolate");
%
 Unstructured??||0
%
 Unstructured??||0
```

But it's hard to tell what we're getting, so let's get the key instead:

```
sqlite> SELECT o.key FROM "_v1_Event" o JOIN "_v1_Event_fields" f on o.key = f.key JOIN "_v1_Event_labels" lt on o.key = lt.key WHERE (lt.label = "bar" and lt.value = "chocolate");
default/my-custom-event4
default/my-custom-event5
```

Unless I'm missing something, this looks ok.  In the next comment I'll report on an object where there are no labels.

Reply:
So here's a `reason="Starting"` event for `metadata.name=lima-rancher-desktop.17e053aa93e89c71` that has no labels.  Here are two curl calls, one without labels, one with, on this event:
```
$ curl -ksL https://localhost:5111/v1/events'?filter=metadata.name="lima-rancher-desktop.17e053aa93e89c71"' | jq . | wc
      94     165    2826
$ curl -ksL https://localhost:5111/v1/events'?filter=metadata.name="lima-rancher-desktop.17e053aa93e89c71"&filter=metadata.labels.fish=car' | jq .
{
  "type": "collection",
  "links": {
    "self": "https://localhost:5111/v1/events"
  },
  "createTypes": {
    "event": "https://localhost:5111/v1/events"
  },
  "actions": {},
  "resourceType": "event",
  "data": []
}
```

The SQL for the second call was this:
```
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
  JOIN "_v1_Event_fields" f ON o.key = f.key
  JOIN "_v1_Event_labels" lt ON o.key = lt.key
  WHERE
    (f."metadata.name" = "lima-rancher-desktop.17e053aa93e89c71") AND
    (lt.label = "fish" AND lt.value LIKE "%car% ESCAPE '\')
  ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC 
  LIMIT 100000
``` 

and it behaved as expected.

Reply:
> My main peeve here is that we're mocking the database in the unit tests, and that makes increasing less sense to me.

100% agreed. DB should be in the loop - we do care about correctness of queries, but much more importantly we care about correctness of results. And those can't be tested without some DB in the loop.

> I've written plenty of unit tests in Rails and Flask (python), and you never mock the database. You just use a test variant of it, and the framework resets the database on every test. Yes, it takes a bit longer, but the tests are much simpler (no mocks) and you have more confidence that the DB code will run correctly in production.

Same experience in my background (Rails and Java/Hibernate).

> So, no, we don't have any tests that cover this.

How about extending pkg/cache/sql/integration_test.go

Those already run a fake Kubernetes environment and have the DB in the loop.

Reply:
> The SQL for the second call was this: 
> 
> ```
> SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
>   JOIN "_v1_Event_fields" f ON o.key = f.key
>   JOIN "_v1_Event_labels" lt ON o.key = lt.key
>   WHERE
>     (f."metadata.name" = "lima-rancher-desktop.17e053aa93e89c71") AND
>     (lt.label = "fish" AND lt.value LIKE "%car% ESCAPE '\')
>   ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC 
>   LIMIT 100000
> ```
> 
> and it behaved as expected.


I tried the same but the behavior I got does not match yours for some reason. (One thing that does not match is the use of double quotes around `"fish"` and `"%car% ESCAPE '\'` - not sure if that is the culprit)


Please try with my DB, obtained by running Rancher briefly with the latest version of this PR: https://drive.google.com/drive/folders/1xIRN5fZ4ye7OD6eHsO-tTOYkOD6RaT5J?usp=sharing


Like in your case, events have no labels:

Query:
```sql
SELECT COUNT(DISTINCT e.key) AS events_with_labels
 FROM _v1_Event e
 WHERE (SELECT COUNT(*) FROM _v1_Event_labels l WHERE e.key = l.key) > 0;
```

Result:
```
0
```

Query:
```sql
SELECT COUNT(DISTINCT e.key) AS events_without_labels
FROM _v1_Event e
WHERE (SELECT COUNT(*) FROM _v1_Event_labels l WHERE e.key = l.key) = 0;
```

Result:
```
228
```

Running the query above (with fixed quotes) returns no results:
```sql
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
                                                 JOIN "_v1_Event_fields" f ON o.key = f.key
                                                 JOIN "_v1_Event_labels" lt ON o.key = lt.key
WHERE
    (f."metadata.name" = "lima-rancher-desktop.17e053aa93e89c71") AND
    (lt.label = "fish" AND lt.value LIKE '%car%' ESCAPE '\')
ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC
  LIMIT 100000;
```

Result:
```
```

And that is correct, as no fish nor car labels exists.

Problem is removing the `WHERE` conditions which is what would happen if there was no filter applied:

```sql
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
                                                 JOIN "_v1_Event_fields" f ON o.key = f.key
                                                 JOIN "_v1_Event_labels" lt ON o.key = lt.key
ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC
  LIMIT 100000;
```

Result:
```
```

That is not correct, as all 228 events should have returned. `JOIN`ing on fields filtered them away, in fact removing the JOIN brings the rows back:

```sql
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
                                                 JOIN "_v1_Event_fields" f ON o.key = f.key
ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC
  LIMIT 3;
```

Result:
```
%
 Unstructured||0
%
 Unstructured||0
%
 Unstructured||0
```


So I continue to think that `JOIN`ing on labels is not what we should do. Filtering on subqueries (something like `WHERE... AND (SELECT COUNT(*) FROM labels l WHERE l.key = o.key) > 0` seems to me to be the best way forward.

Reply:
First, correcting the query (for anyone who wants to follow along at home):
```
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
  JOIN "_v1_Event_fields" f ON o.key = f.key
  JOIN "_v1_Event_labels" lt ON o.key = lt.key
  WHERE
    (f."metadata.name" = "lima-rancher-desktop.17e053aa93e89c71") AND
    (lt.label = "fish" AND lt.value LIKE "%car%" ESCAPE '\')
  ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC 
  LIMIT 100000
```

And this does return no rows, as expected.

Reply:
Next, regarding this query:
```
SELECT o.object, o.objectnonce, o.dekid FROM "_v1_Event" o
                                                 JOIN "_v1_Event_fields" f ON o.key = f.key
                                                 JOIN "_v1_Event_labels" lt ON o.key = lt.key
ORDER BY f."metadata.namespace" ASC, f."metadata.name" ASC
  LIMIT 100000;
  ```

I only do a join on `X_labels` if there's an actual label query going on. This isn't the case,
so this query should never be executed. If you can provide a selector string that results in
this query, please let me know and I'll add it to the unit tests for `constructQuery`.

Reply:
What about a != label filter though?

https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#equality-based-requirement says:

```
environment = production
tier != frontend
```

> The former selects all resources with key equal to environment and value equal to production. The latter selects all resources with `key` equal to `tier` and value distinct from `frontend`, **and all resources with no labels with the `tier` key.**

(emphasis mine)

Reply:
Actually, I implemented `NOT EXISTS` for labels incorrectly, and didn't catch that second case for the `labels != value` form, so they're both wrong.